Source/Utils/MsgLogger/MsgLogger.H

```diff
@@ -0,0 +1,364 @@
+    };
+
+    /**
+    * \brief This implements the < operator for Msg.
+    * Warning messages are first ordered by priority (warning: high < medium < low
+    * to give precedence to higher priorities), then by topic (alphabetically),
+    * and finally by text (alphabetically).
+    *
+    * @param[in] l a Msg
+    * @param[in] r a Msg
+    * @return true if l<r, false otherwise
+    */
+    constexpr bool operator<(const Msg& l, const Msg& r)
+    {
+        return
+            (l.priority > r.priority) ||
+            ((l.priority == r.priority) && (l.topic < r.topic)) ||
+            ((l.priority == r.priority) && (l.topic == r.topic) && (l.text < r.text));
+    }
+
+    /**
+    * This class is responsible for storing messages and merging messages
+    * collected by different processes.
+    */
+    class Logger
+    {
+        public:
+
+        /**
+        * \brief The constructor.
+        */
+        Logger();
+
+        /**
+        * \brief This function records a message
+        *
+        * @param[in] msg a Msg struct
+        */
+        void record_msg(Msg msg);
+
+        /**
+        * \brief This function returns a vector containing the recorded messages
+        *
+        * @return a vector of the recorded messages
+        */
+        std::vector<Msg> get_msgs() const;
+
+        /**
+        * \brief This function returns a vector containing the recorded messages
+        * with the corresponding counters
+        *
+        * @return a vector of the recorded messages with counters
+        */
+        std::vector<MsgWithCounter> get_msgs_with_counter() const;
+
+        /**
+        * \brief This collective function generates a vector containing the messages
+        * with counters and emitting ranks by gathering data from
+        * all the ranks
+        *
+        * @return a vector of messages with counters and ranks if I/O rank, an empty vector otherwise
+        */
+        std::vector<MsgWithCounterAndRanks>
+        collective_gather_msgs_with_counter_and_ranks() const;
+
+        private:
+
+        /**
+        * \brief This function implements the trivial special case of
+        * collective_gather_msgs_with_counter_and_ranks when there is only one rank.
+        *
+        * @return a vector of messages with counters and ranks
+        */
+        std::vector<MsgWithCounterAndRanks>
+        one_rank_gather_msgs_with_counter_and_ranks() const;
+
+#ifdef AMREX_USE_MPI
+        /**
+        * \brief This collective function finds the rank having the
+        * most messages and how many messages this rank has. The
+        * rank having the most messages is designated as "gather rank".
+        *
+        * @param[in] how_many_msgs the number of messages that the current rank has
+        * @return a pair containing the ID of the "gather rank" and its number of messages
+        */
+        std::pair<int,int> find_gather_rank_and_its_msgs(
+            int how_many_msgs) const;
+
+        /**
```

Comment:
After discussing with @NeilZaim , we think that these static private functions, which are essentially helper functions for `collective_gather_msgs_with_counter_and_ranks` could be moved to the `.cpp` file (similarly to what suggested here: https://stackoverflow.com/questions/59318867/where-to-put-helper-functions-pertaining-to-class-method-in-c)

Reply:
This has just been implemented.